libs-base/Tools/cvtenc.1
.\"Copyright (C) 2005 Free Software Foundation, Inc.
.\"Copying and distribution of this file, with or without modification,
.\"are permitted in any medium without royalty provided the copyright
.\"notice and this notice are preserved.
.TH CVTENC "1" "February 2004" "GNUstep" "GNUstep System Manual"
.SH NAME
cvtenc \- file encoding converter

.SH SYNOPSIS
.B cvtenc
[\fB-Encoding\fR \fIencoding\fR]
[\fB-EscapeIn\fR \fIyes|NO\fR]
[\fB-EscapeOut\fR \fIyes|NO\fR]
 \fIfilename\fR

.SH DESCRIPTION
.P
Converts a file encoded in a specified or default non-unicode encoding to
unicode (UTF16), or, if the file is already in unicode, converts it to a
specified or default non-unicode encoding.
The converted text is printed to standard out.
If the filename is a hyphen then this tool reads from standard input.

.SH OPTIONS
.IP "\fB\-Encoding\fR \fI encoding\fR" 4
Specify non-unicode encoding to be used.  The direction of conversion depends
on whether the file given as input is recognized as unicode (from the higher
order bytes).  To get a list of available encodings, enter "list" for the
encoding.  If this is not specified, the default encoding for the current
locale is used.
.IP "\fB\-EscapeIn\fR \fIYES|NO\fR" 4
Specify '\-EscapeIn YES' (the default is 'NO') to parse the input for \\u
escape sequences (as in property lists).
.IP "\fB\-EscapeOut\fR \fIYES|NO\fR" 4
Specify '\-EscapeOut YES' (the default is 'NO') to generate \\u escape
sequences (as in property lists) in the output.  Note, this might produce
unexpected results for some encodings.
.IP "\fB\-Unicode\fR \fIIN|OUT\fR" 4
Specify '\-Unicode IN' (or '\-Unicode OUT') to control the direction of
conversion rather than having the tool guess it on the basis of the content
it reads.  Using '\-Unicode IN' means that the tool reads UTF16 data and
writes C-String data, while using '\-Unicode OUT' merans that the tool reads
C-String data and writes UTF16 data.


.SH HISTORY
.RS 0
Written in 2002.
Minr update in 2010.
.P
This manual page first appeared in gnustep-base 1.9.2 (March 2004).
.P
.SH AUTHORS
.B cvtenc
was written by Richard Frith-Macdonald <rfm@gnu.org>


libs-gui/NEWS
1 NEWS
******

The currently released version of the library is '0.28.0'.

1.1 Noteworthy changes in version '0.28.0'
==========================================

This version adds support for modern XIB files and many new classes.
Plus the usual bunch of bug fixes.

   * Support loading of document XIB files.
   * Improve Key Value Binding for NSArrayController and add more
     bindings.
   * Better support for multi monitor usage and other improvement in the
     backend integration.
   * Add classes NSFontCollection, NSColorSampler, NSSpeechRecognizer,
     NSAppearance, NSPDFInfo, NSPICTImageRep, NSCIImageRep,
     NSPDFImageRep, NSPDFPanel, NSDataAsset, NSDatePicker,
     NSDatePickerCell, NSPredicateEditor, NSPredicateEditorRowTemplate,
     NSRuleEditor, NSGestureRecognizer, NSButtonTouchBarItem,
     NSCandidateListTouchBarItem, NSClickGestureRecognizer,
     NSColorPickerTouchBarItem, NSCustomTouchBarItem,
     NSGroupTouchBarItem, NSMagnificationGestureRecognizer,
     NSPanGestureRecognizer, NSPickerTouchBarItem,
     NSPopoverTouchBarItem, NSPressGestureRecognizer,
     NSRotationGestureRecognizer, NSSharingServicePickerTouchBarItem,
     NSSliderTouchBarItem, NSStepperTouchBarItem, NSTouchBarItem,
     NSTouchBar, NSTouch, NSDockTile.
   * Implement NSEPSImageRep.
   * Better encoding handling in RTF files.
   * Theming and drawing improvements.
   * Increase small font size to 10.
   * New cursor and stepper images.
   * Move NSFileWrapper to Foundation.
   * Fixed build on Debian GNU/kFreeBSD.
   * With command line argument -autolaunch YES, do not activate the
     application when -activateIgnoringOtherApps: is invoked.
   * Improvements to WindowMaker compatibility (e.g.  WMFHideApplication
     support).
   * Lowered NSFloatingWindowLevel by one to distinguish floating panels
     from menus.

1.2 Noteworthy changes in version '0.27.0'
==========================================

This version includes numerous bugfixes, compatibility improvements and
other changes accumulated over the last year.  It also enables work to
be done on integrating NSViews with a Core Animation renderer.

   * Make targetForAction safer.
   * Speed up menu updates.
   * Clean up speech tool compilation and switch to newer interface.
   * Fix bug in CUPS subclassing introduced in last release.
   * Minor improvements to typesetting.
   * Add NSIsControllerMarker.
   * Fix tracking on segmented cell.
   * Bring slider cell closer to Cocoa implementation.
   * Add ivar for Core Animation in NSView.
   * Improve border calculation on printing.
   * Lazy load app icon.
   * Better detection of removable volumes.
   * Polish translations.
   * Japanese translations.
   * Lots of bug fixes.

1.3 Noteworthy changes in version '0.26.2'
==========================================

This version is a small, but important bugfix release.

   * printing: Fix allocation of the CUPS printing classes.
   * installation: Fix the configure script.

1.4 Noteworthy changes in version '0.26.1'
==========================================

This version is released to conincide with version 1.25.1 of
gnustep-base, which contains changes required for this version of
gnustep-gui and gnustep-back.

   It includes an important workaround for users of GNUstep Objective-C
Runtime (libobjc2) and non-fragile ABI to avoid a bug in interaction
between the clang compiler and the runtime when non-fragile ABI is in
use.  Specifically, Clang and the runtime may disagree on what is the
offset of an ivar in a class's RAM. This manifested in a crash at
application startup due to misalignment of _gcontext inside NSThread.
See the mailing list discussion
(http://lists.gnu.org/archive/html/discuss-gnustep/2017-12/msg00129.html)
for more information.

   It also contains the following changes:

   * tests: Cleanup of warnings.
   * tests: Fix text system deallocation test.
   * printing: Undefine __BLOCKS__ before including cups.h, as some
     versions of the header expect that libdispatch is present and used
     if __BLOCKS__ is defined.
   * graphics context: Workaround for Clang+libobjc2+nonfragile ABI
     issue.

1.5 Noteworthy changes in version '0.26.0'
==========================================

This version was bumped due to previous binary incompatibilities between
0.25.0 and 0.25.1.  This version also includes numerous compatibility
improvements from the Summer of Code project, and a wide variety of
other fixes.  Notably, it fixes the use of cupsGetPPD() in the printing
system.

   * printing: Add an include to get deprecated function cupsGetPPD() on
     newer CUPS systems.
   * chore: Bump required base version.
   * tiff: Support for writing resolution.
   * jpeg: Save resolution information if it is different from 72 dpi.
   * save panel: Fix return type of sorting function.
   * events: Add some newer Cocoa enums and one method with dummy
     implementation.
   * speech synthesis: NSSpeechSynthesizerDelegate is now a @protocol on
     runtimes that support it.
   * pasteboard: New type identifiers.
   * translations: Some work on Polish, Russian and German translations
   * cell: Improvements to mouse tracking logic on NSCell.
   * image: If an unknown named image is unarchived with a coder or
     keyed coder, keep the name.
   * screen: Add -backingScaleFactor and return 1.0.
   * window: Return 1.0 from -backingScaleFactor.

   * compatibility: Numerous stub implementations of constants, classes
     and methods to improve source-level compatibility.
   * other bugfixes

1.6 Noteworthy changes in version '0.25.1'
==========================================

   * JPEG (saving) alpha channel fixes and size with resolution != 72
   * JPEG resolution read support
   * TIFF saving fixes
   * Improved volumes mounting and support
   * Portability improvements in volume mounting and support
   * Corrected layout of empty strings
   * Only update visible menus

1.7 Noteworthy changes in version '0.25.0'

libs-base/Tools/cvtenc.m
/** This tool converts a file containing a string to a C String encoding.
   Copyright (C) 2002 Free Software Foundation, Inc.

   Written by:  Richard Frith-Macdonald <richard@brainstorm.co.uk>
   Created: April 2002

   This file is part of the GNUstep Project

   This program is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License
   as published by the Free Software Foundation; either
   version 3 of the License, or (at your option) any later version.

   You should have received a copy of the GNU General Public
   License along with this program; see the file COPYINGv3.
   If not, write to the Free Software Foundation,
   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.

   */

#import "common.h"

#import	"Foundation/NSArray.h"
#import	"Foundation/NSData.h"
#import	"Foundation/NSException.h"
#import	"Foundation/NSString.h"
#import	"Foundation/NSProcessInfo.h"
#import	"Foundation/NSUserDefaults.h"
#import	"Foundation/NSFileHandle.h"
#import	"Foundation/NSAutoreleasePool.h"
#import "GNUstepBase/NSString+GNUstepBase.h"
#import "GNUstepBase/GSMime.h"
#ifdef NeXT_Foundation_LIBRARY
#import "GNUstepBase/Additions.h"
#endif

#include	<ctype.h>

/** Return whether value ch between min and max. */
#define inrange(ch,min,max) ((ch)>=(min) && (ch)<=(max))
/** Convert hex digit in ascii to decimal equivalent. */
#define char2num(ch) \
inrange(ch,'0','9') \
? ((ch)-0x30) \
: (inrange(ch,'a','f') \
? ((ch)-0x57) : ((ch)-0x37))


/** <p>Converts  a file encoded in a specified or default non-unicode encoding
 *  to unicode, or, if the file is already in unicode,  converts  it  to  a
 *  specified  or  default  non-unicode  encoding.   The  converted text is
 *  printed to standard out.</p>
 */
int
main(int argc, char** argv, char **env)
{
  NSAutoreleasePool	*pool;
  NSProcessInfo		*proc;
  NSArray		*args;
  unsigned		i;
  BOOL			found = NO;
  BOOL			eIn;
  BOOL			eOut;
  NSString		*n;
  NSStringEncoding	enc = 0;

#ifdef GS_PASS_ARGUMENTS
  GSInitializeProcess(argc, argv, env);
#endif
  pool = [NSAutoreleasePool new];
  proc = [NSProcessInfo processInfo];
  if (proc == nil)
    {
      NSLog(@"defaults: unable to get process information!\n");
      [pool release];
      exit(EXIT_SUCCESS);
    }

  args = [proc arguments];

  eIn = [[NSUserDefaults standardUserDefaults] boolForKey: @"EscapeIn"];
  eOut = [[NSUserDefaults standardUserDefaults] boolForKey: @"EscapeOut"];
  n = [[NSUserDefaults standardUserDefaults] stringForKey: @"Encoding"];
  if (n == nil)
    {
      enc = [NSString defaultCStringEncoding];
    }
  else if (0 == (enc = [GSMimeDocument encodingFromCharset: n]))
    {
      const NSStringEncoding	*e;
      NSMutableString	*names;

      names = [NSMutableString stringWithCapacity: 1024];
      e = [NSString availableStringEncodings];
      while (*e != 0)
	{
	  NSString	*name = [NSString localizedNameOfStringEncoding: *e];

	  [names appendFormat: @"  '%@'\n", name];
	  if ([n isEqual: name] == YES)
	    {
	      enc = *e;
	      break;
	    }
	  e++;
	}
      if (enc == 0)
	{
	  NSLog(@"defaults: unable to find encoding '%@'!\n"
	    @"Localised encoding names are -\n%@", n, names);
	  [pool release];
	  exit(EXIT_SUCCESS);
	}
    }

  n = [[NSUserDefaults standardUserDefaults] stringForKey: @"Unicode"];
  n = [[n stringByTrimmingSpaces] lowercaseString];
  if ([n length] > 0)
    {
      if ([n isEqual: @"in"] || [n isEqual: @"i"])
	{
	  n = @"i";
	}
      else if ([n isEqual: @"out"] || [n isEqual: @"o"])
	{
	  n = @"o";
	}
      else
	{
	  n = nil;
	}
    }
  else
    {
      n = nil;
    }

  for (i = 1; found == NO && i < [args count]; i++)
    {
      NSString	*file = [args objectAtIndex: i];

      if ([file hasPrefix: @"-"] == YES && NO == [file isEqual: @"-"])
	{
	  i++;
	  continue;
	}
      found = YES;
      NS_DURING
	{
	  NSData	*myData;

          if (YES == [file isEqual: @"-"])
	    {
	      myData = [[[NSFileHandle fileHandleWithStandardInput]
		readDataToEndOfFile] retain];
	    }
	  else
	    {
	      myData = [[NSData alloc] initWithContentsOfFile: file];
	    }
	  if (myData == nil)
	    {
	      NSLog(@"File read operation failed for %@.", file);
	    }
	  else
	    {
	      unsigned		l = [myData length];
	      const unichar	*b = (const unichar*)[myData bytes];
	      NSStringEncoding	iEnc;
	      NSStringEncoding	oEnc;
	      NSString		*myString;

	      if (nil == n)
		{
		  if (l > 1 && (*b == 0xFFFE || *b == 0xFEFF))
		    {
		      iEnc = NSUnicodeStringEncoding;
		      oEnc = enc;
		    }
		  else
		    {
		      iEnc = enc;
		      oEnc = NSUnicodeStringEncoding;
		    }
		}
	      else if ([n isEqualToString: @"i"])
		{
		  /* Unicode (UTF16) in
		   */
		  iEnc = NSUnicodeStringEncoding;
		  oEnc = enc;
		}
	      else
		{
		  /* Unicode (UTF16) out
		   */
		  iEnc = enc;
		  oEnc = NSUnicodeStringEncoding;
		}
	
	      myString = [[NSString alloc] initWithData: myData
					       encoding: iEnc];
	      RELEASE(myData);
	      if (myString == nil)
		{
		  NSLog(@"Encoding input conversion failed for %@.", file);
		}
	      else
		{
		  if (eIn == YES)
		    {
		      unsigned	l = [myString length];
		      unichar	*u;
		      NSZone	*z = NSDefaultMallocZone();
		      unsigned	i = 0;
		      unsigned	o = 0;

		      u = NSZoneMalloc(z, sizeof(unichar)*l);
		      [myString getCharacters: u];

		      while (i < l)
			{
			  unichar	c = u[i++];

			  if (c == '\\' && i <= l - 6)
			    {
			      c = u[i++];

			      if (c == 'u' || c == 'U')
				{
				  unichar	v;

				  v = 0;
				  c = u[i++];
				  v |= char2num(c);

				  v <<= 4;
				  c = u[i++];
				  v |= char2num(c);
				
				  v <<= 4;
				  c = u[i++];
				  v |= char2num(c);
				
				  v <<= 4;
				  c = u[i++];
				  v |= char2num(c);

				  c = v;
				}
			      else
				{
				  u[o++] = '\\';
				}
			    }
			  u[o++] = c;
			}

		      RELEASE(myString);
		      myString = [[NSString alloc] initWithCharactersNoCopy: u
			length: o freeWhenDone: YES];
		    }
		  if (eOut == YES)
		    {
		      unsigned	l = [myString length];
		      unichar	*u;
		      char	*c;
		      NSZone	*z = NSDefaultMallocZone();
		      unsigned	o = 0;
		      unsigned	i;

		      u = NSZoneMalloc(z, sizeof(unichar)*l);
		      c = NSZoneMalloc(z, 6*l);
		      [myString getCharacters: u];
		      for (i = 0; i < l; i++)
			{
			  if (u[i] < 128)
			    {
			      c[o++] = u[i];
			    }
			  else
			    {
                              unsigned  v = u[i];

                              c[o++] = '\\';
                              c[o++] = 'U';
                              c[3] = "0123456789abcdef"[v & 0xf];
                              v /= 16;
                              c[2] = "0123456789abcdef"[v & 0xf];
                              v /= 16;
                              c[1] = "0123456789abcdef"[v & 0xf];
                              v /= 16;
                              c[0] = "0123456789abcdef"[v & 0xf];
			      o += 4;
			    }
			}
		      NSZoneFree(z, u);
		      myData = [[[NSData alloc]
			initWithBytesNoCopy: c length: o] autorelease];
		    }
		  else if (eIn == YES)
		    {
		      myData = [myString dataUsingEncoding: iEnc
				      allowLossyConversion: NO];
		    }
		  else
		    {
		      myData = [myString dataUsingEncoding: oEnc
				      allowLossyConversion: NO];
		    }
		  RELEASE(myString);
		  if (myData == nil)
		    {
		      NSLog(@"Encoding output conversion failed for %@.",
			file);
		    }
		  else
		    {
		      NSFileHandle	*out;

		      ENTER_POOL
		      out = [NSFileHandle fileHandleWithStandardOutput];
		      [out writeData: myData];
		      [out synchronizeFile];
		      LEAVE_POOL
		    }
		}
	    }
	}
      NS_HANDLER
	{
	  NSLog(@"Converting '%@' - %@", file, [localException reason]);
	}
      NS_ENDHANDLER
    }

  if (found == NO)
    {
      NSLog(@"\nThis utility expects a filename as an argument.\n"
	@"It reads the file, and writes it to STDOUT after converting it\n"
	@"to unicode (UTF16) from C-string encoding or vice versa.\n"
	@"You can use '-' as the filename argument to read from STDIN.\n"
	@"You can supply a '-Encoding name' option to specify the C-string\n"
	@"encoding to be used, if you don't want to use the default.\n"
	@"If you supply an unknown encoding the tool will print a list\n"
	@"of all the known encodings.\n"
	@"You can supply a '-EscapeIn YES' option to specify that input\n"
	@"should be parsed for \\U escape sequences (as in property lists).\n"
	@"You can supply a '-EscapeOut YES' option to specify that output\n"
	@"should be ascii with \\U escape sequences (for property lists).\n"
	@"You can supply a '-Unicode in/out' option to specify that the\n"
	@"conversion is from/to unicode (UTF16).  This suppresses the normal\n"
	@"behavior of guessing the direction of conversion from the content\n"
	@"of the incoming data.\n");
    }
  [pool release];
  return 0;
}
